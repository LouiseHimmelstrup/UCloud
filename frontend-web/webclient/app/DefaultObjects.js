import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};