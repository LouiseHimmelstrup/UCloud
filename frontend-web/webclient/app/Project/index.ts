import {APICallParameters} from "Authentication/DataHook";
import {buildQueryString} from "Utilities/URIUtilities";
import {Page, PaginationRequest} from "Types";
import {Client} from "Authentication/HttpClientInstance";
import {DEV_SITE, STAGING_SITE} from "../../site.config.json";
import {inDevEnvironment} from "UtilityFunctions";
import {IconName} from "ui-components/Icon";

const groupContext = "/projects/groups/";
const projectContext = "/projects/";

interface CreateGroupRequest {
    group: string;
}

export interface ListGroupMembersRequestProps extends PaginationRequest {
    group: string;
}

export function createGroup(props: CreateGroupRequest): APICallParameters<{}> {
    return {
        reloadId: Math.random(),
        method: "PUT",
        path: groupContext,
        parameters: props,
    };
}

export function listGroupMembersRequest(
    props: ListGroupMembersRequestProps
): APICallParameters<ListGroupMembersRequestProps> {
    return {
        method: "GET",
        path: buildQueryString(`${groupContext}members`, props),
        reloadId: Math.random(),
        parameters: props
    };
}

interface AddGroupMemberProps {
    group: string;
    memberUsername: string;
}

export function addGroupMember(payload: AddGroupMemberProps): APICallParameters<AddGroupMemberProps> {
    return {
        method: "PUT",
        path: `${groupContext}/members`,
        reloadId: Math.random(),
        payload
    };
}

interface RemoveGroupMemberProps {
    group: string;
    memberUsername: string;
}

export function removeGroupMemberRequest(payload: RemoveGroupMemberProps): APICallParameters<RemoveGroupMemberProps> {
    return {
        method: "DELETE",
        path: `${groupContext}members`,
        reloadId: Math.random(),
        payload
    };
}

export interface ShouldVerifyMembershipResponse {
    shouldVerify: boolean;
}

export function shouldVerifyMembership(projectId: string): APICallParameters {
    return {
        method: "GET",
        path: `${projectContext}should-verify`,
        reloadId: Math.random(),
        projectOverride: projectId
    };
}

export function verifyMembership(projectId: string): APICallParameters {
    return {
        method: "POST",
        path: `${projectContext}verify-membership`,
        reloadId: Math.random(),
        projectOverride: projectId
    };
}

export function projectRoleToString(role: ProjectRole): string {
    switch (role) {
        case ProjectRole.PI: return "PI";
        case ProjectRole.ADMIN: return "Admin";
        case ProjectRole.USER: return "User";
    }
}

export function projectStringToRole(role: string): ProjectRole {
    switch (role) {
        case "PI": return ProjectRole.PI;
        case "Admin": return ProjectRole.ADMIN;
        case "User": return ProjectRole.USER;
    }
    return ProjectRole.USER;
}

export function projectRoleToStringIcon(role: ProjectRole): IconName {
    switch (role) {
        case ProjectRole.PI: return "userPi";
        case ProjectRole.ADMIN: return "userAdmin";
        case ProjectRole.USER: return "user";
    }
}

export function groupSummaryRequest(payload: PaginationRequest): APICallParameters<PaginationRequest> {
    return {
        path: buildQueryString(`${groupContext}/summary`, payload),
        method: "GET",
        reloadId: Math.random(),
        parameters: payload,
        payload
    };
}

export interface UserStatusRequest {

}

export interface UserStatusResponse {
    membership: UserInProject[];
    groups: UserGroupSummary[];
}

export function userProjectStatus(request: UserStatusRequest): APICallParameters<UserStatusRequest> {
    return {
        reloadId: Math.random(),
        path: "/projects/membership",
        method: "POST",
        parameters: request,
        payload: request
    };
}

export interface MembershipSearchRequest extends PaginationRequest {
    query: string;
    notInGroup?: string;
}

type MembershipResponse = Page<ProjectMember>;

export function membershipSearch(request: MembershipSearchRequest): APICallParameters<MembershipSearchRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/membership/search", request),
        parameters: request
    };
}

export interface UpdateGroupNameRequest {
    oldGroupName: string;
    newGroupName: string;
}

export function updateGroupName(request: UpdateGroupNameRequest): APICallParameters<UpdateGroupNameRequest> {
    return {
        method: "POST",
        path: "/projects/groups/update-name",
        parameters: request,
        payload: request
    };
}

export interface DeleteGroupRequest {
    groups: string[];
}

export function deleteGroup(request: DeleteGroupRequest): APICallParameters<DeleteGroupRequest> {
    return {
        method: "DELETE",
        path: "/projects/groups",
        parameters: request,
        payload: request
    };
}

export interface ProjectMember {
    username: string;
    role: ProjectRole;
}

export interface Project {
    id: string;
    title: string;
    members: ProjectMember[];
}

export const emptyProject = (id: string): Project => ({
    id,
    title: "",
    members: []
});

export enum ProjectRole {
    PI = "PI",
    ADMIN = "ADMIN",
    USER = "USER"
}

export interface UserInProject {
    projectId: string;
    title: string;
    whoami: ProjectMember;
    needsVerification: boolean;
    favorite: boolean;
    archived: boolean;
}

export interface UserGroupSummary {
    projectId: string;
    group: string;
    username: string;
}

export const createProject = (payload: {title: string}): APICallParameters => ({
    method: "POST",
    path: "/projects",
    payload,
    reloadId: Math.random()
});

export const inviteMember = (payload: {projectId: string; usernames: string[]}): APICallParameters => ({
    method: "POST",
    path: "/projects/invites",
    payload,
    reloadId: Math.random()
});

export const deleteMemberInProject = (payload: {projectId: string; member: string}): APICallParameters => ({
    method: "DELETE",
    path: "/projects/members",
    payload,
    reloadId: Math.random()
});

export const changeRoleInProject = (
    payload: {projectId: string; member: string; newRole: ProjectRole}
): APICallParameters => ({
    method: "POST",
    path: "/projects/members/change-role",
    payload,
    reloadId: Math.random()
});

export interface ListProjectsRequest extends PaginationRequest{
    archived?: boolean;
    noFavorites?: boolean;
}

export const listProjects = (parameters: ListProjectsRequest): APICallParameters<ListProjectsRequest> => ({
    method: "GET",
    path: buildQueryString(
        "/projects/list",
        parameters
    ),
    parameters,
    reloadId: Math.random()
});

export interface ListFavoriteProjectsRequest extends PaginationRequest {
    archived: boolean;
}

export const listFavoriteProjects = (parameters: ListFavoriteProjectsRequest): APICallParameters<ListFavoriteProjectsRequest> => ({
    method: "GET",
    path: buildQueryString(
        "/projects/listFavorites",
        parameters
    ),
    parameters,
    reloadId: Math.random()
});

export const roleInProject = (project: ProjectMember[]): ProjectRole | undefined => {
    const member = project.find(m => {
        return m.username === Client.username;
    });

    if (member === undefined) return undefined;
    return member.role;
};


export interface ToggleProjectFavorite{
    projectId: string;
}

export function toggleFavoriteProject(request: ToggleProjectFavorite): APICallParameters<ToggleProjectFavorite> {
    return {
        method: "POST",
        path: "/projects/favorite",
        parameters: request,
        payload: request
    };
}

export interface OutgoingInvite {
    username: string;
    invitedBy: string;
    timestamp: number;
}

export interface ListOutgoingInvitesRequest extends PaginationRequest {

}

export function listOutgoingInvites(request: ListOutgoingInvitesRequest): APICallParameters<ListOutgoingInvitesRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/invites/outgoing", request),
        parameters: request
    };
}

export interface IngoingInvite {
    project: string;
    invitedBy: string;
    timestamp: string;
}

export interface ListIngoingInvitesRequest extends PaginationRequest {

}

export function listIngoingInvites(request: ListIngoingInvitesRequest): APICallParameters<ListIngoingInvitesRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/invites/ingoing", request),
        parameters: request
    };
}

export interface AcceptInviteRequest {
    projectId: string;
}

export function acceptInvite(request: AcceptInviteRequest): APICallParameters<AcceptInviteRequest> {
    return {
        method: "POST",
        path: "/projects/invites/accept",
        payload: request,
        parameters: request
    };
}

export interface RejectInviteRequest {
    projectId: string;
    username?: string;
}

export function rejectInvite(request: RejectInviteRequest): APICallParameters<RejectInviteRequest> {
    return {
        method: "DELETE",
        path: "/projects/invites/reject",
        parameters: request,
        payload: request
    };
}

type LeaveProjectRequest = {};
export function leaveProject(request: LeaveProjectRequest): APICallParameters<LeaveProjectRequest> {
    return {
        method: "DELETE",
        path: "/projects/leave",
        parameters: request,
        payload: request
    };
}

export interface TransferPiRoleRequest {
    newPrincipalInvestigator: string;
}

export function transferPiRole(request: TransferPiRoleRequest): APICallParameters<TransferPiRoleRequest> {
    return {
        method: "POST",
        path: "/projects/transfer-pi",
        parameters: request,
        payload: request
    };
}

export type ListRepositoryRequest = PaginationRequest;

export function listRepositoryFiles(
    request: ListRepositoryRequest,
    projectOverride?: string
): APICallParameters<ListRepositoryRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/repositories/list-files", request),
        parameters: request,
        reloadId: Math.random(),
        projectOverride
    };
}

export interface ArchiveProjectRequest {
    archiveStatus: boolean;
}

export function setProjectArchiveStatus(request: ArchiveProjectRequest): APICallParameters<ArchiveProjectRequest> {
    return {
        method: "POST",
        path: "/projects/archive",
        parameters: request,
        payload: request,
        reloadId: Math.random()
    };
}

export interface ViewProjectRequest {
    id: string;
}

export function viewProject(request: ViewProjectRequest): APICallParameters<ViewProjectRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/view", request),
        parameters: request,
        reloadId: Math.random()
    };
}

export function areProjectsEnabled(): boolean {
    if ([DEV_SITE, STAGING_SITE].includes(window.location.host) || inDevEnvironment()) {
        return true;
    }
    return Client.userRole === "ADMIN";
}
