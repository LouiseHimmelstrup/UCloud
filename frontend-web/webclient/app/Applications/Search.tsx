import * as React from "react";
import {Box, Stamp} from "@/ui-components";
import {emptyPage} from "@/DefaultObjects";
import {joinToString} from "@/UtilityFunctions";
import {useLocation, useNavigate} from "react-router";
import {useEffect} from "react";
import {buildQueryString, getQueryParamOrElse} from "@/Utilities/URIUtilities";
import {useCloudAPI, useCloudCommand} from "@/Authentication/DataHook";
import * as UCloud from "@/UCloud";
import {GridCardGroup} from "@/ui-components/Grid";
import {ApplicationCard} from "@/Applications/Card";
import * as Pagination from "@/Pagination";
import {FilesSearchTabs} from "@/Files/FilesSearchTabs";

interface SearchQuery {
    tags: string[];
    query: string;
    showAllVersions: boolean;
    page: number;
    itemsPerPage: number;
}

function readQuery(queryParams: string): SearchQuery {
    const tags: string[] = [];
    const tagsQuery = getQueryParamOrElse(queryParams, "tags", "");
    tagsQuery.split(",").forEach(it => {
        if (it !== "") {
            tags.push(it);
        }
    });
    const showAllVersions = getQueryParamOrElse(queryParams, "showAllVersions", "false") === "true";
    const query = getQueryParamOrElse(queryParams, "query", "");
    let itemsPerPage = parseInt(getQueryParamOrElse(queryParams, "itemsPerPage", "25"), 10);
    let page = parseInt(getQueryParamOrElse(queryParams, "page", "0"), 10);
    if (isNaN(itemsPerPage) || itemsPerPage <= 0) itemsPerPage = 25;
    if (isNaN(page) || page < 0) page = 0;

    return {query, tags, showAllVersions, itemsPerPage, page};
}

export const SearchResults: React.FunctionComponent<{entriesPerPage: number}> = ({entriesPerPage}) => {
    const navigate = useNavigate();
    const location = useLocation();
    const [, invokeCommand] = useCloudCommand();
    const [results, fetchResults] = useCloudAPI<UCloud.Page<UCloud.compute.ApplicationSummaryWithFavorite>>(
        {noop: true},
        emptyPage
    );

    const queryParams = location.search;
    const parsedQuery = readQuery(queryParams);

    useEffect(() => {
        fetchResults(
            UCloud.compute.apps.searchApps({
                query: new URLSearchParams(queryParams).get("q") ?? "",
                itemsPerPage: 100,
                page: parsedQuery.page,
            })
        );
    }, [queryParams, /* This isn't needed for this one, is it? */entriesPerPage]);

    const toggleFavorite = React.useCallback(async (appName: string, appVersion: string) => {
        await invokeCommand(UCloud.compute.apps.toggleFavorite({appName, appVersion}));
        fetchResults(
            UCloud.compute.apps.searchApps({
                query: new URLSearchParams(queryParams).get("q") ?? "",
                itemsPerPage: 100,
                page: 0
            })
        );
    }, [fetch]);

    return <>
        <FilesSearchTabs active={"APPLICATIONS"} />
        <Pagination.List
            loading={results.loading}
            page={results.data}
            pageRenderer={page => (
                <GridCardGroup>
                    {page.items.map(app => (
                        <ApplicationCard
                            key={`${app.metadata.name}${app.metadata.version}`}
                            app={app}
                            onFavorite={toggleFavorite}
                            isFavorite={app.favorite}
                            tags={app.tags}
                        />))
                    }
                </GridCardGroup>
            )}
            onPageChanged={newPage => {
                navigate(buildQueryString("/applications/search", {
                    q: new URLSearchParams(queryParams).get("q") ?? "",
                    tags: joinToString(parsedQuery.tags),
                    showAllVersions: parsedQuery.showAllVersions.toString(),
                    page: newPage.toString(),
                    itemsPerPage: parsedQuery.itemsPerPage.toString(),
                }));
            }}
        />
    </>;
};
