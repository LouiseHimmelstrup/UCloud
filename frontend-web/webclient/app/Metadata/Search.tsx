import * as React from "react";
import { Link } from "react-router-dom";
import { simpleSearch, ProjectMetadata } from "./api";
import { Page } from "Types";
import { emptyPage } from "DefaultObjects";
import { withRouter } from "react-router-dom";
import * as PropTypes from "prop-types";
import * as Pagination from "Pagination";
import { KeyCode } from "DefaultObjects";
import { History } from "history";
import { updatePageTitle } from "Navigation/Redux/StatusActions";
import * as Heading from "ui-components/Heading";
import { Hide, Input, Box, Flex, Text, Stamp, Divider, Card } from "ui-components";

interface SearchState {
    query: string
    dataProvider: (page: number, itemsPerPage: number) => Promise<Page<ProjectMetadata>>
}

interface SearchProps {
    history: History
    match: {
        params: {
            query?: string
        }
    }
}

class SearchComponent extends React.Component<SearchProps, SearchState> {
    constructor(props, ctx) {
        super(props);
        this.state = {
            query: "",
            dataProvider: (page: number, itemsPerPage: number) => Promise.resolve(emptyPage)
        };
        ctx.store.dispatch(updatePageTitle("Search"));
    }

    static contextTypes = {
        store: PropTypes.object
    }

    componentDidMount() {
        this.checkQueryParams();
    }

    componentDidUpdate() {
        // TODO Doesn't work correctly with react router on back/forwards browser navigation
        this.checkQueryParams();
    }

    private checkQueryParams() {
        const query = this.props.match.params.query;
        if (query == null) {
            return;
        } else if (this.state.query != query) {
            this.setState(() => ({
                query,
                dataProvider: (page: number, itemsPerPage: number) => simpleSearch(query.toLowerCase(), page, itemsPerPage)
            }));
        }
    }

    render() {
        const { history } = this.props;
        return (
            <div>
                <Hide xl lg md>
                    <Input
                        width="100%"
                        placeholder="Search..."
                        onKeyDown={(e: any) => { if (e.keyCode === KeyCode.ENTER) history.push(`/metadata/search/${e.target.value}`) }}
                    />
                </Hide>
                <Heading.h2>Results matching '{this.state.query}'</Heading.h2>

                <Pagination.ManagedList
                    dataProvider={this.state.dataProvider}
                    pageRenderer={this.pageRenderer}
                />
            </div>
        );
    }

    pageRenderer = (page: Page<ProjectMetadata>): React.ReactNode => (
        <>{page.items.map((item, index) => <SearchItem item={item} key={index} />)}</>
    )
}

export const SearchItem = ({ item }: { item: ProjectMetadata }) => (
    <Card height="154px" p="12px" mb="0.5em" mt="0.5em" borderRadius=".28571429rem">
        <Heading.h3><Link to={`/metadata/${item.sduCloudRoot}`}>{item.title}</Link></Heading.h3>

        <Divider />
        <Box mt="1em" mb="1em">
            <Text>
                {firstParagraphWithLimitedLength(
                    defaultIfBlank(item.description, "No description"),
                    800
                )}
            </Text>
        </Box>
        <Divider mb="1em" />
        <Stamp bg="green" color="white" borderColor="green">
            <Box pl="0.5em" pr="0.5em">
                <i className="fas fa-folder-open" />
            </Box>
            Open Access
            </Stamp>
        <Stamp bg="blue" color="white" borderColor="blue" ml="0.2em">
            <Flex>
                <Box pl="0.5em" pr="0.5em">
                    <i className="fas fa-book"></i>
                </Box>
                MIT
                <Text pl="0.2em" color="lightGrey">License</Text>
            </Flex>
        </Stamp>
    </Card>
);

const defaultIfBlank = (text: string, defaultValue: string): string => {
    if (text.length == 0) return defaultValue;
    else return text;
};

const firstParagraphWithLimitedLength = (text: string, maxLength: number): string => {
    const lines = text.split("\n");
    const paragraphEndsAt = lines.findIndex((line) => /^\s*$/.test(line));

    let firstParagraph: string;
    if (paragraphEndsAt == -1) firstParagraph = text;
    else firstParagraph = lines.slice(0, paragraphEndsAt).join("\n");

    if (firstParagraph.length > maxLength) {
        return firstParagraph.substring(0, maxLength) + "...";
    } else {
        return firstParagraph;
    }
};

export const Search = withRouter(SearchComponent);