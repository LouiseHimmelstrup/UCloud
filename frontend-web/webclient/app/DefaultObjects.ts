import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState} from "Files";
import {SortOrder, SortBy} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Publication} from "Zenodo";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import * as SnackbarRedux from "Snackbar/Redux";
import * as FavoritesRedux from "Favorites/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum AnalysesStatusMap {
    "PENDING",
    "IN PROGRESS",
    "COMPLETED"
}

export enum RightsNameMap {
    "NONE" = "None",
    "READ" = "Read",
    "READ_WRITE" = "Read/Write"
}

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

function getFilesSortingColumnOrDefault(columnIndex: 0 | 1): SortBy {
    const sortingColumn = window.localStorage.getItem(`filesSorting${columnIndex}`);
    if (sortingColumn && Object.values(SortBy).includes(sortingColumn)) return sortingColumn as SortBy;
    switch (columnIndex) {
        case 0:
            window.localStorage.setItem("filesSorting0", SortBy.MODIFIED_AT);
            return SortBy.MODIFIED_AT;
        case 1:
            window.localStorage.setItem("filesSorting1", SortBy.SIZE);
            return SortBy.SIZE;
    }
}

function getItemOrDefault<T, T2>(itemName: string, defaultValue: T, en: T2): T {
    const item = window.localStorage.getItem(itemName);
    if (item && Object.values(en).includes(item)) {
        return item as unknown as T;
    }
    return defaultValue;
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FilesReduxObject extends ComponentWithPage<File> {
    sortOrder: SortOrder
    sortBy: SortBy
    path: string
    filesInfoPath: string
    fileSelectorError?: string
    sortingColumns: [SortBy, SortBy]
    fileSelectorLoading: boolean
    fileSelectorShown: boolean
    fileSelectorPage: Page<File>
    fileSelectorPath: string
    fileSelectorCallback: Function
    fileSelectorIsFavorites: boolean
    disallowedPaths: string[]
    invalidPath: boolean
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface ZenodoReduxObject extends ComponentWithPage<Publication> {
    connected: boolean
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    files?: Reducer<FilesReduxObject>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    zenodo?: Reducer<ZenodoReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
    detailedResult?: Reducer<DetailedResultReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

export type DetailedResultReduxObject = ComponentWithPage<File>

export const initDetailedResult = (): DetailedResultReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined
});

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    files: FilesReduxObject,
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    zenodo: ZenodoReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    detailedResult: DetailedResultReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects &
    FavoritesRedux.Objects &
    SnackbarRedux.Wrapper;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        files: initFiles(homeFolder),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        zenodo: initZenodo(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        detailedResult: initDetailedResult(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        ...FavoritesRedux.init(),
        ...SnackbarRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.ASCENDING
});

export const initZenodo = (): ZenodoReduxObject => ({
    connected: false,
    loading: false,
    page: emptyPage
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFiles = (homeFolder: string): FilesReduxObject => ({
    page: emptyPage,
    sortOrder: getItemOrDefault("sortOrder", SortOrder.ASCENDING, SortOrder),
    sortBy: getItemOrDefault("sortBy", SortBy.PATH, SortBy),
    loading: false,
    error: undefined,
    path: "",
    filesInfoPath: "",
    sortingColumns: [getFilesSortingColumnOrDefault(0), getFilesSortingColumnOrDefault(1)],
    fileSelectorLoading: false,
    fileSelectorShown: false,
    fileSelectorPage: emptyPage,
    fileSelectorPath: homeFolder,
    fileSelectorIsFavorites: false,
    fileSelectorCallback: () => undefined,
    fileSelectorError: undefined,
    disallowedPaths: [],
    invalidPath: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
