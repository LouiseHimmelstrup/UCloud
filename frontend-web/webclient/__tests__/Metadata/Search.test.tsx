import * as React from "react";
import { Search } from "Project/Search";
import { MemoryRouter } from "react-router";
import { configure, mount } from "enzyme";
import * as PropTypes from "prop-types";
import * as Adapter from "enzyme-adapter-react-16";
import { configureStore } from "Utilities/ReduxUtilities";
import { initStatus } from "DefaultObjects";
import status from "Navigation/Redux/StatusReducer";
import { Provider } from "react-redux";

configure({ adapter: new Adapter() });

describe("Search component", () => {
    test("Mount search component", () => {
        const store = configureStore({ status: initStatus() }, { status });
        expect(mount(<MemoryRouter><Search /></MemoryRouter>, { context: { store: store }, childContextTypes: { store: PropTypes.object} }).html()).toMatchSnapshot()
    })
});