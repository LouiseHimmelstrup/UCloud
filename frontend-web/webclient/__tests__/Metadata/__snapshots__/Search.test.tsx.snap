// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component Mount search component 1`] = `"<div><div class=\\"sc-kEYyzF bkuwjF\\"><input width=\\"100%\\" placeholder=\\"Search...\\" id=\\"default\\" class=\\"sc-iAyFgw hOTwqy\\"></div><h2 font-size=\\"5\\" class=\\"sc-jzJRlG fXjJRp\\">Results matching ''</h2><div class=\\"sc-bZQynM cNFsWw\\"><div class=\\"sc-htpNat hjOqln\\"></div><div class=\\"sc-Rmtcm idAtlS\\"><div class=\\"sc-bdVaJa hfvaUK\\"><span cursor=\\"pointer\\" class=\\"sc-htoDjs bChhWZ\\"><div class=\\"sc-bZQynM bqIUrb\\"><div class=\\"sc-htpNat llOEfe\\"> Items per page 25</div><div class=\\"sc-htpNat llOEfe\\"><svg cursor=\\"auto\\" class=\\"sc-VigVT hlFKbn sc-gqjmRU eNpPdi\\" viewBox=\\"0 0 24 24\\" width=\\"24\\" height=\\"24\\" fill=\\"currentcolor\\"><path d=\\"M7.4,8l4.6,4.6L16.6,8L18,9.4l-6,6l-6-6L7.4,8z\\"></path></svg></div></div></span></div><i class=\\"fas fa-sync fa-spin\\"></i></div></div><div class=\\"sc-bZQynM bqIUrb\\"><div width=\\"50%\\" class=\\"sc-htpNat knGhlu\\"></div><i class=\\"fas fa-circle-notch fa-spin\\"></i></div><div></div></div>"`;
